tests/unit/test_oauth_requirer.py

```diff
@@ -0,0 +1,176 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import pytest
+from charms.hydra.v0.oauth import (
+    CLIENT_SECRET_FIELD,
+    ClientConfig,
+    ClientConfigError,
+    InvalidClientConfigEvent,
+    OAuthInfoChangedEvent,
+    OAuthRequirer,
+    _load_data,
+)
+from ops.charm import CharmBase
+from ops.testing import Harness
+
+METADATA = """
+name: requirer-tester
+requires:
+  oauth:
+    interface: oauth
+"""
+
+
+@pytest.fixture()
+def harness():
+    harness = Harness(OAuthRequirerCharm, meta=METADATA)
+    harness.set_leader(True)
+    harness.begin_with_initial_hooks()
+    yield harness
+    harness.cleanup()
+
+
+@pytest.fixture()
+def provider_info():
+    return {
+        "authorization_endpoint": "https://example.oidc.com/oauth2/auth",
+        "introspection_endpoint": "https://example.oidc.com/admin/oauth2/introspect",
+        "issuer_url": "https://example.oidc.com",
+        "jwks_endpoint": "https://example.oidc.com/.well-known/jwks.json",
+        "scope": "openid profile email phone",
+        "token_endpoint": "https://example.oidc.com/oauth2/token",
+        "userinfo_endpoint": "https://example.oidc.com/userinfo",
+    }
+
+
+CLIENT_CONFIG = {
+    "redirect_uri": "https://example.oidc.client/callback",
+    "scope": "openid email offline_access",
+    "grant_types": ["authorization_code", "refresh_token"],
+    "audience": [],
+    "token_endpoint_auth_method": "client_secret_basic",
+}
+
+
+class OAuthRequirerCharm(CharmBase):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args)
+        client_config = ClientConfig(**CLIENT_CONFIG)
+        self.oauth = OAuthRequirer(self, client_config=client_config)
+
+        self.events = []
+        self.framework.observe(self.oauth.on.oauth_info_changed, self._record_event)
+        self.framework.observe(self.oauth.on.invalid_client_config, self._record_event)
+
+    def _record_event(self, event):
+        self.events.append(event)
+
+
+def test_data_in_relation_bag_on_joined(harness):
+    relation_id = harness.add_relation("oauth", "provider")
+
+    relation_data = harness.get_relation_data(relation_id, harness.model.app.name)
+
+    assert _load_data(relation_data) == CLIENT_CONFIG
+
+
+def test_oauth_info_changed_emitted_on_client_creation(harness, provider_info):
```

Comment:
This unit test is very long. Ideally, the goal is for a test to only have 1 reason to crash. While it's fine to have more than 1 `assert` (ex. if there are multiple values you want to check in a dict), you really only want to validate 1 behavior.

Reply:
I split it into 2 tests. 
Currently, I am checking that the correct event is emitted, but also that the `client_secret_id` in the event corresponds to the correct juju secret, is that ok with you or do you think I should try to break it down even further?